### PR TITLE
Fix typings

### LINKED ISSUE #3752: Types definition for Typescript
## Feature Request

**Is your feature request related to a problem? Please describe.**

Yes, there is no good type definition for this library. All the types are just `any`.

**Describe the suggestion or request in detail**

Please generate appropriate type definition for Typescript.

**Additional context**

image

### LINKED ISSUE #3856: Unable to import createDirectLine, createStore in Typescript with v4.13
## Screenshots

file `lib\index.d.ts` from version v4.13.0 (left) vs v4.11.0 (right)

image

## Version

NPM: v4.13.0

## Describe the bug

The issue is present in the npm package `botframework-webchat` used in a react website (`https://github.com/microsoft/BotFramework-WebChat/tree/master/packages/bundle/src`)

The output lib folder contains a `index.d.ts` that is created using the new `src\index.tsx` and not `src\index.ts`. This is probably an issue with webpack output configuration in v4.13

Due to this the following line fails to compile in ts/vscode
```
import ReactWebChat, { createDirectLine, createStore } from 'botframework-webchat';
```
image

explicitly referencing lib/index.js does not work and is unable to recognize the exports.

Considering this work in v4.11.0, I would consider this a regression introduced in v4.13

## Steps to reproduce

1. Create react app using create react app
1. Add package botframework-webchat v 4.13.0
1. Import createDirectLine, createStore from 'botframework-webchat';
1. See error

## Expected behavior

The bundled package should support importing createDirectLine, createStore from 'botframework-webchat'

The above code works in v4.11.0. 

## Additional context

Suggest to rename the `src\index.tsx` added as part of PR #3818, creating a correct webpack output
The react samples in `https://github.com/microsoft/BotFramework-WebChat/tree/master/samples/01.getting-started/g.hybrid-react-npm/packages/chat-component` have not been updated to v4.13 else they would showcase similar issues.

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #3856. Fixes #3752.

### Added

-  Resolves [#3752](https://github.com/microsoft/BotFramework-WebChat/issues/3752). Added typings (`*.d.ts`) for all public interfaces, by [@compulim](https://github.com), in PR [#3931](https://github.com/microsoft/BotFramework-WebChat/pull/3931)

### Fixed

-  Fixes [#3856](https://github.com/microsoft/BotFramework-WebChat/issues/3856). Fix missing typings, by [@compulim](https://github.com/compulim) and [@corinagum](https://github.com/corinagum), in PR [#3931](https://github.com/microsoft/BotFramework-WebChat/pull/3931)

## Description

When we work on #3818, some typings were missing. This PR is to add typings to all public interfaces.

Although they are now properly typed, we still need to add human-readable description to each interface. For example, copying documentations from [`HOOKS.md`](https://github.com/microsoft/BotFramework-WebChat/blob/master/docs/HOOKS.md) to each `use*.js` (see #3932).

In the future, we should unify the public interface of middleware. For example, attachment middleware use this interface:

```js
() => () => () => React.Element
```

While newer scroll-to-end button middleware use this interface:

```js
() => () => () => false | ((props) => React.Element)
```

The newer interface allow caller to know if the middleware decided to render something, or want to skip it. This is useful for transcript, as we can know "how many activities are going to render" beforehand, and this help grouping activities.

### Updated test hierarchy

Test assets under `/__tests__/types/__typescript__/pass` are TypeScript code that mimic how web developers use Web Chat.

Test assets under `/__tests__/types/__typescript__/fail-once` are TypeScript code that should fail because Web Chat public interfaces was used in a wrong way (wrong types, no children allowed, etc.)

### `<AddFullBundle>`

To remove duplicate code between `FullReactWebChat.tsx` and `FullComposer.tsx`, we added `AddFullBundle.tsx` to refactor their functionalities.

The following is copied from the updated `FullReactWebChat.tsx` and it has much less code than previous versions.

```jsx
const FullReactWebChat: VFC<FullReactWebChatProps> = props => (
  <AddFullBundle {...props}>{extraProps => <ReactWebChat {...props} {...extraProps} />}</AddFullBundle>
);
```

### Deprecated or deprecate-to-be are not typed

We are planning to deprecate Redux store when we work on ACS. Thus, the store state is not typed.

Interfaces which are already deprecated, are not typed.

## Specific Changes

- Update all public interfaces, from `.js` to `.ts`

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] ~I have updated documentation~

## Review Checklist

> This section is for contributors to review your work.

-  [x] ~Accessibility reviewed (tab order, content readability, alt text, color contrast)~
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] ~CSS styles reviewed (minimal rules, no `z-index`)~
-  [x] ~Documents reviewed (docs, samples, live demo)~
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] `package.json` and `package-lock.json` reviewed
-  [x] ~Security reviewed (no data URIs, check for nonce leak)~
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: packages/api/src/utils/createCustomEvent.ts [hunk 1]
```diff
@@ -13,5 +18,5 @@ export default function createCustomEvent(name, eventInitDict) {
     event[key] = value;
   });
 
-  return event;
+  return event as any;
```
Comment (cwhitten) [inline packages/api/src/utils/createCustomEvent.ts:21]: you type event to `Event` and then return as `any`.. Intended?

File: packages/bundle/src/adaptiveCards/Attachment/ReceiptCardContent.tsx [hunk 2]
```diff
@@ -47,32 +56,54 @@ const ReceiptCardContent = ({ actionPerformedClassName, content, disabled }) =>
...
+
+              itemColumns = columns;
+              builder.addImage(url, itemImageColumn, imageTap, alt);
+            } else {
+              itemColumns = builder.addColumnSet([75, 25], undefined, tap && tap);
+            }
+
+            const [itemTitleColumn, itemPriceColumn] = itemColumns;
+
+            builder.addTextBlock(
```
Comment (cwhitten) [inline packages/bundle/src/adaptiveCards/Attachment/ReceiptCardContent.tsx:94]: some of this looks like new implementation. why is it needed?

File: packages/bundle/src/adaptiveCards/createAdaptiveCardsAttachmentMiddleware.tsx [hunk 3]
```diff
@@ -10,30 +12,30 @@ import SignInCardAttachment from './Attachment/SignInCardAttachment';
...
+export default function createAdaptiveCardsAttachmentMiddleware(): AttachmentMiddleware {
   // This is not returning a React component, but a render function.
   /* eslint-disable-next-line react/display-name */
   return () => next => (...args) => {
-    const [{ activity, attachment }] = args;
+    const [{ attachment }] = args;
 
     return attachment.contentType === 'application/vnd.microsoft.card.hero' ? (
-      <HeroCardAttachment activity={activity} attachment={attachment} />
+      <HeroCardAttachment attachment={attachment} />
```
Comment (cwhitten) [inline packages/bundle/src/adaptiveCards/createAdaptiveCardsAttachmentMiddleware.tsx:22]: Where is the activity coming from in these components? Redux?

File: packages/bundle/src/tsconfig.json [hunk 4]
```diff
@@ -11,5 +11,5 @@
     "skipLibCheck": true,
     "sourceMap": true
   },
-  "files": ["index.tsx", "index-es5.tsx", "index-minimal.tsx"]
```
Comment (cwhitten) [inline packages/bundle/src/tsconfig.json:14]: why is this change needed?

File: packages/component/src/Attachment/VimeoContent.tsx [hunk 5]
```diff
@@ -1,18 +1,25 @@
...
   const [{ vimeoContent: vimeoContentStyleSet }] = useStyleSet();
 
   const search = new URLSearchParams({
-    autoplay: autoPlay ? 1 : 0,
-    badge: 0,
-    byline: 0,
-    loop: loop ? 1 : 0,
-    portrait: 0,
-    title: 0
+    autoplay: autoPlay ? '1' : '0',
```
Comment (cwhitten) [inline packages/component/src/Attachment/VimeoContent.tsx:17]: why the need to change this from a number to a string that looks like a number?

File: packages/component/src/SendBox/SuggestedActions.tsx [hunk 6]
```diff
@@ -223,7 +229,7 @@ const SuggestedActions = ({ className, suggestedActions = [] }) => {
 
   const children = suggestedActions.map(({ displayText, image, imageAltText, text, title, type, value }, index) => (
     <SuggestedAction
-      ariaHidden={true}
+      aria-hidden={true}
```
Comment (cwhitten) [inline packages/component/src/SendBox/SuggestedActions.tsx:232]: why change this?

File: packages/component/src/SendBox/TextBox.tsx [hunk 7]
```diff
@@ -409,7 +419,7 @@ const TextBox = ({ className }) => {
           placeholder={typeYourMessageString}
           readOnly={disabled}
           ref={inputElementRef}
-          rows="1"
+          rows={1}
```
Comment (cwhitten) [inline packages/component/src/SendBox/TextBox.tsx:422]: Why change this?

**Review by cwhitten — COMMENTED** [review]
First pass done

**Comment by compulim** [comment]
Ready for second pass.

**Review by cwhitten — APPROVED** [review]
